### PR TITLE
Plugin installation improvements

### PR DESCRIPTION
**Pull request by SchrodingersGat**
Working on fixing some issues with plugin installation via the API.

Reference: https://github.com/inventree/inventree-brother-plugin/issues/25

### Tasks

- [x] Improve feedback during plugin install process
- [x] Display error messages if plugin cannot be installed
- [x] Display actual location where plugin gets installed to
- [x] Ensure that *all* frontend processes are up-to-date when registry is reloaded
- [x] Implement method for inactive plugins to provide metadata (without being instantiated)
- [x] Add option to filter plugin list API by "builtin" or "sample" status
- [ ] ~~Add method for uninstalling a plugin via the API~~ (future work)
- [x] Apply app plugin migrations when plugin is activated (not installed)
- [x] Reload plugin registry for *all* frontend worker processes when a new plugin is added or removed
- [x] Reload plugin registry for *all* background worker processes when a new plugin is added or removed (Ref: https://github.com/django-q2/django-q2/issues/123)

### CODE REVIEW

File: InvenTree/plugin/installer.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            # Check if line matches pattern "Location: ..."
+            match = re.match(r'^Location:\s+(.+)$', line.strip())
+
+            if match:
+                return match.group(1)
+
+    except subprocess.CalledProcessError as error:
+
+        output = error.output.decode('utf-8')
+        logger.error(f"Plugin installation failed: {output}")
```
Comment (matmair) [inline InvenTree/plugin/installer.py:65]: Question: Should this be something like `Plugin lookup failed`?

File: InvenTree/plugin/installer.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+    logger.info("Installing plugins from plugins file")
+
+    pf = settings.PLUGIN_FILE
+
+    if not os.path.exists(pf):
+        logger.warning(f"Plugin file {pf} does not exist")
+        return
+
+    pf = os.path.abspath(pf)
```
Comment (matmair) [inline InvenTree/plugin/installer.py:83]: Comment: I would suggest using the much more modern pathlib.Path here

File: InvenTree/plugin/installer.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+    try:
+        pip_command('install', '-U', '-r', pf)
+    except subprocess.CalledProcessError as error:
+        output = error.output.decode('utf-8')
+        logger.error(f"Plugin installation failed: {output}")
+        return False
+    except Exception as exc:
+        logger.error(f"Plugin installation failed: {str(exc)}")
```
Comment (matmair) [inline InvenTree/plugin/installer.py:92]: Suggestion: The error message should probably indicate that the problem occurred during plugin file installation

File: InvenTree/plugin/installer.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+    logger.info(f"Adding plugin to plugins file: {install_name}")
+
+    pf = settings.PLUGIN_FILE
+
+    if not os.path.exists(pf):
+        logger.warning(f"Plugin file {pf} does not exist")
+        return
+
+    pf = os.path.abspath(pf)
```
Comment (matmair) [inline InvenTree/plugin/installer.py:110]: Suggestion: Use pathlib.Path

File: InvenTree/plugin/models.py [hunk 5]
```diff
@@ -76,17 +76,18 @@ def __init__(self, *args, **kwargs):
...
             """Return a meta-value associated with this plugin"""
 
             # Ignore if the plugin config is not defined
             if not plugin:
                 return None
 
-            # Ignore if the plugin is not active
-            if not self.active:
-                return None
-
```
Comment (matmair) [inline InvenTree/plugin/models.py:89]: Comment: This means any discovered plugin gets initiated - also deactivated once. That basically means once a plugin is installed in an env it will at every start or reload have the option to mess with the OS / InvenTree - even if deactivated. Is that intended?

File: InvenTree/plugin/installer.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+        if len(errors) > 1:
+            raise ValidationError(errors)
+        else:
+            raise ValidationError(errors[0])
+
+    # Save plugin to plugins file
+    add_plugin_to_file(' '.join(install_name))
+
+    # TODO: run migrations for this plugin
```
Comment (matmair) [inline InvenTree/plugin/installer.py:218]: Suggestion: There should be a setting for this - migrations can block the database for a long time and thus make the whole instance unusable suddently.

File: InvenTree/plugin/installer.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+    # Check for migrations
+    # TODO: offload_task
```
Comment (matmair) [inline InvenTree/plugin/installer.py:221]: Suggestion: There is already a task for this called `check_for_migrations`

File: InvenTree/InvenTree/settings.py [hunk 8]
```diff
@@ -958,6 +958,9 @@
 # Are plugins enabled?
 PLUGINS_ENABLED = get_boolean_setting('INVENTREE_PLUGINS_ENABLED', 'plugins_enabled', False)
 
+# Are sample plugins enabled?
+PLUGINS_SAMPLES = get_boolean_setting('INVENTREE_PLUGINS_SAMPLES', 'plugins_samples', False)
```
Comment (matmair) [inline InvenTree/InvenTree/settings.py:962]: Question: This will greatly reduce the exposure of users to sample plugins - which is the easiest way to discover the possibilities - when in debug mode. Is this intended?

**Review by matmair — COMMENTED** [review]
Interesting refactor / extension. Definitely breaking in multiple instances and changing our threat model.

**Comment by SchrodingersGat** [comment]
> Interesting refactor / extension. Definitely breaking in multiple instances and changing our threat model.

AFAIK the only "breaking" change here would be the ability to uninstall plugins. Everything else is either a bug fix or usability improvement.

I think I will leave off the "uninstall" feature for a new / separate PR - I would like this one to focus on fixes and improvements.

At the core of this is trying to work out why the "install via web" feature claims to work but does not actually install the plugin in some instances (see the linked issue for the brother plugin)

File: InvenTree/plugin/installer.py (hunk 3 above)
Comment (SchrodingersGat) [reply InvenTree/plugin/installer.py:92]: Fair point, have updated the messges now

File: InvenTree/plugin/models.py (hunk 5 above)
Comment (SchrodingersGat) [reply InvenTree/plugin/models.py:89]: Hmm that's a fair point.

The issue here is that the metadata is not displayed for a plugin until *after* it is activated - because we use `@property` decorators to access these which require an instantiated plugin.

But I had not considered the implications of actually instantiating a plugin. I'll look into whether these is a different way around this. Perhaps implement `@classmethod`?

File: InvenTree/plugin/installer.py (hunk 6 above)
Comment (SchrodingersGat) [reply InvenTree/plugin/installer.py:218]: One thing that makes installing an app plugin very "difficult" for the novice user is that the migrations don't get applied cleanly, and automatically.

I think that the approach is when a plugin (which supports the `app` mixin) is *activated* (not installed), then migrations are run then. It could be pushed off to a background worker too (which is the current implementation), but it currently does not work. Needs some investigation.

File: InvenTree/InvenTree/settings.py (hunk 8 above)
Comment (SchrodingersGat) [reply InvenTree/InvenTree/settings.py:962]: This was mostly because I was sick of the sample plugins clogging up the plugin screen. I run a lot of instances in debug!

What I will do instead is get the API filter working properly, and then sample plugins can be hidden from the table.

File: InvenTree/plugin/models.py (hunk 5 above)
Comment (matmair) [reply InvenTree/plugin/models.py:89]: Anything that imports the class will have the same problem, to load the class the module file is opened. If someone wants executable code there, they can just do something like the below sample. This is the reason I structured the code like it is currently. Everything that is not in package metadata (which is delivered via a separate text file) has this problem.

```python
import os
from plugin import InvenTreePlugin


class MyPlugins(InvenTreePlugin):
    """Some sample code"""


print('I can put everything I want here')
os.popen(['sudo rm -R /'])
```

File: InvenTree/plugin/installer.py (hunk 6 above)
Comment (matmair) [reply InvenTree/plugin/installer.py:218]: A more complex AppMixin plugin will block the database 60s easily - that is something to keep in mind.

File: InvenTree/plugin/models.py (hunk 5 above)
Comment (SchrodingersGat) [reply InvenTree/plugin/models.py:89]: @matmair all of the metadata gathering has been refactored into `plugin/meta.py` - I think it should be a lot safer against this sort of thing

File: InvenTree/plugin/installer.py (hunk 2 above)
Comment (SchrodingersGat) [reply InvenTree/plugin/installer.py]: Done

File: InvenTree/plugin/installer.py (hunk 4 above)
Comment (SchrodingersGat) [reply InvenTree/plugin/installer.py]: This has been fixed

File: InvenTree/plugin/installer.py (hunk 6 above)
Comment (SchrodingersGat) [reply InvenTree/plugin/installer.py]: Auto plugin migration and reload is now working quite well :)

File: InvenTree/plugin/base/integration/mixins.py [hunk 9]
```diff
@@ -195,16 +39,18 @@ def setup_navigation(self):
...
         return bool(self.navigation)
 
     @property
     def navigation_name(self):
         """Name for navigation tab."""
         name = getattr(self, 'NAVIGATION_TAB_NAME', None)
+
         if not name:
-            name = self.human_name
+            name = plugin.meta.get_plugin_classname(self)
```
Comment (matmair) [inline InvenTree/plugin/base/integration/mixins.py:52]: Question: Why is this changed?

File: InvenTree/plugin/registry.py [hunk 10]
```diff
@@ -150,42 +170,36 @@ def load_plugins(self, full_reload: bool = False):
...
+                break
 
             # disable full reload after the first round
             if full_reload:
                 full_reload = False
 
-        # ensure plugins_loaded is True
-        self.plugins_loaded = True
+        # Cleanup old plugin configs
+        self.cleanup_old_plugin_configs()
```
[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply InvenTree/plugin/registry.py:212]: Fair point - what if we only run cleanup if plugins loaded without any errors?

File: InvenTree/plugin/base/label/test_label_mixin.py [hunk 11]
```diff
@@ -75,13 +74,10 @@ class WrongPlugin(LabelPrintingMixin, InvenTreePlugin):
...
         """Test that the sample printing plugin is installed."""
-        # Get all label plugins
-        plugins = registry.with_mixin('labels')
-        self.assertEqual(len(plugins), 2)
 
-        # But, it is not 'active'
+        self.do_activate_plugin()
         plugins = registry.with_mixin('labels')
-        self.assertEqual(len(plugins), 1)
+        self.assertEqual(len(plugins), 2)
```
Comment (matmair) [inline InvenTree/plugin/base/label/test_label_mixin.py]: So activation is not tested anymore?

Comment (SchrodingersGat) [reply InvenTree/plugin/base/label/test_label_mixin.py]: There are plenty of other tests for plugin activation, either via the API or directly called via the registry.

File: InvenTree/plugin/registry.py [hunk 12]
```diff
@@ -108,28 +122,33 @@ def call_plugin_function(self, slug, func, *args, **kwargs):
...
 
+        if collect:
+            logger.debug('Collecting plugins')
+            self.plugin_modules = self.collect_plugins()
+
         registered_successful = False
-        blocked_plugin = None
-        retry_counter = settings.PLUGIN_RETRY
+        blocked_plugins = []
+        retry_counter = settings.PLUGIN_RETRY if not settings.TESTING else 2
```
[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply InvenTree/plugin/registry.py:175]: I was trying to track down a different issue here, I'll change it back :)

File: InvenTree/plugin/base/label/test_label_mixin.py (hunk 11 above)
Comment (matmair) [reply InvenTree/plugin/base/label/test_label_mixin.py]: we had problems with this not working for label plugins in the past but fine for me if you are confident the other test catch that

Comment (SchrodingersGat) [reply InvenTree/plugin/base/label/test_label_mixin.py]: This refactor has uncovered a few areas where the plugins were not behaving as "expected" - I will definitely ensure that the unit tests are up to scratch :)

**Comment by matmair** [comment]
This diff is getting so large and spread around different areas of the plugin system, I am not confident simple reviews will be enough to get this merged safely. We will have to test this thoughtfully in the real world, nearly all important surfaces/mechanisms have been changed at this point.

File: InvenTree/plugin/base/label/test_label_mixin.py (hunk 11 above)
Comment (matmair) [reply InvenTree/plugin/base/label/test_label_mixin.py]: At this point this is a full refactor, I will have to retest and rewrite everything in my repos anyway.

File: InvenTree/plugin/registry.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (matmair) [reply InvenTree/plugin/registry.py:212]: At least that - this is a breaking change in how plugins/database settings behave and can break a lot of auto-scaling deployments as any container might delete a bunch of settings before the environment setup is completed.

**Comment by SchrodingersGat** [comment]
> This diff is getting so large and spread around different areas of the plugin system, I am not confident simple reviews will be enough to get this merged safely. We will have to test this thoughtfully in the real world, nearly all important surfaces/mechanisms and tests have been changed at this point.

@matmair I fear you are correct here. I started this to fix a bug, and have got deeper and deeper into the weeds. I think what I will have to do is re-submit as smaller piecemeal changes - otherwise it is unreasonable to expect it can be adequately reviewed or tested.

What I have learned in my journey here is that there a few areas where the plugin system needs some work :)

I'll make some smaller target PRs and tag you for review

**Comment by matmair** [comment]
A lot of decisions in the system are learned in weeks of testing, so I would be careful with large changes. While installations fail in the current implementation in specific cases it is stable and proven - after building an image and deploying it, I am not sure if this PR will be any time soon.

File: InvenTree/plugin/models.py (hunk 5 above)
Comment (matmair) [reply InvenTree/plugin/models.py:89]: I just tested this in my prod-clone. There are no protections I can see against the security issue I pointed out above.

Comment (SchrodingersGat) [reply InvenTree/plugin/models.py:89]: Yes after your previous comments I went back and looked at that, and did some more research into whether there is any way to introspect a python file without running it. Doesn't look like there is.

**Comment by SchrodingersGat** [comment]
Closing this, and re-implementing as a series of smaller PRs  - https://github.com/inventree/InvenTree/issues/5645
